onnx/reference/ops/op_conv.py

```diff
@@ -1,11 +1,178 @@
 # SPDX-License-Identifier: Apache-2.0
 # pylint: disable=R0912,R0913,R0914,R0915,R1702,W0221
 
+from typing import List, Tuple
+
 import numpy as np
 
 from onnx.reference.op_run import OpRun
 
 
+def _get_indices(i, shape):  # type: ignore
+    res = np.empty((len(shape),), dtype=np.int64)
+    k = len(shape) - 1
+    while k > 0:
+        m = i % shape[k]
+        res[k] = m
+        i -= m
+        i /= shape[k]
+        k -= 1
+    res[0] = i
+    return res
+
+
+def _is_out(ind, shape):  # type: ignore
+    for i, s in zip(ind, shape):
+        if i < 0:
+            return True
+        if i >= s:
+            return True
+    return False
+
+
+def im2col_naive_implementation(data, kernel_shape, dilations, pads, strides):  # type: ignore
+    """
+    Naive implementation for `im2col`.
+
+    :param image: image (float)
+    :param kernel_shape: kernel shape
+    :param pads: pads
+    :return: result
+    """
+    if not isinstance(kernel_shape, tuple):
+        raise TypeError(f"Unexpected type {type(kernel_shape)!r} for kernel_shape.")
+    if len(data.shape) != len(kernel_shape):
+        raise ValueError(f"Shape mismatch {data.shape!r} and {kernel_shape!r}.")
+    n_dims = len(pads) // 2
+    new_pads = np.array([(pads[i], pads[i + n_dims]) for i in range(n_dims)])
+    list_output_shape = list(data.shape + kernel_shape)
+    for d in range(n_dims):
+        kd = kernel_shape[d] + (kernel_shape[d] - 1) * (dilations[d] - 1)
+        nd = int(
+            ((list_output_shape[d] - kd + new_pads[d][0] + new_pads[d][1]) / strides[d])
+            + 1
+        )
+        list_output_shape[d] = nd
+    output_shape = tuple(list_output_shape)
+
+    res = np.zeros(output_shape, dtype=data.dtype)
+    kernel_size = np.prod(kernel_shape)
+    res_size = np.prod(res.shape[:-n_dims])
+    for i in range(res_size):
+        i_res = _get_indices(i, res.shape[:-n_dims])
+        t_res = tuple(i_res)
+        for j in range(kernel_size):
+            i_kernel = _get_indices(j, kernel_shape)
+            t_kernel = tuple(i_kernel)
+
+            i_img = i_res * strides - new_pads[:, 0] + i_kernel * dilations
+            t_img = tuple(i_img)
+            if _is_out(t_img, data.shape):
+                res[t_res + t_kernel] = 0
+            else:
+                res[t_res + t_kernel] = data[tuple(t_img)]
+    return res
+
+
+def im2col(
+    img: np.ndarray,
+    kernel_shape: Tuple[int, ...],
+    dilations: List[int],
+    pads: List[int],
```

Comment:
Use Sequence when possible to allow more valid inputs?

Reply:
done